PublicGitArchive/pga/cmd/cache.go

```diff
@@ -16,46 +15,68 @@ const (
 	indexName = "latest.csv.gz"
 )
 
-// copy checks whether a new version of the file in url exists and downloads it
-// to dest. It returns true when the file exists, and an error when it was not possible
-// to update it.
-func copy(dest, source FileSystem, name string) (bool, error) {
+// updateCache checks whether a new version of the file in url exists and downloads it
+// to dest. It returns an error when it was not possible to update it.
+func updateCache(dest, source FileSystem, name string) error {
 	logrus.Debugf("syncing %s to %s", source.Abs(name), dest.Abs(name))
-	localTime, err := dest.ModTime(name)
-	exists := !os.IsNotExist(err)
-	if err != nil && !os.IsNotExist(err) {
-		return false, fmt.Errorf("could not check mod time in %s: %v", dest.Abs(name), err)
-	}
-
-	remoteTime, err := source.ModTime(name)
-	if err != nil {
-		return exists, fmt.Errorf("could not check mod time in %s: %v", source.Abs(name), err)
-	}
-
-	if !localTime.IsZero() && !remoteTime.IsZero() && remoteTime.Before(localTime) {
+	if upToDate(dest, source, name) {
 		logrus.Debugf("local copy is up to date")
-		return true, nil
+		return nil
 	}
-	logrus.Debugf("local copy is outdated or non existent")
 
+	logrus.Debugf("local copy is outdated or non existent")
 	wc, err := dest.Create(name)
 	if err != nil {
-		return false, fmt.Errorf("could not create %s: %v", dest.Abs(name), err)
+		return fmt.Errorf("could not create %s: %v", dest.Abs(name), err)
 	}
 
 	rc, err := source.Open(name)
 	if err != nil {
-		return false, err
+		return err
 	}
 	defer rc.Close()
 
 	if _, err = io.Copy(wc, rc); err != nil {
-		return false, fmt.Errorf("could not copy %s to %s: %v", source.Abs(name), dest.Abs(name), err)
+		return fmt.Errorf("could not copy %s to %s: %v", source.Abs(name), dest.Abs(name), err)
 	}
 	if err := wc.Close(); err != nil {
-		return false, fmt.Errorf("could not close %s: %v", dest.Abs(name), err)
+		return fmt.Errorf("could not close %s: %v", dest.Abs(name), err)
+	}
+	return nil
+}
+
+func upToDate(dest, source FileSystem, name string) bool {
+	ok, err := matchHash(dest, source, name)
+	if err == nil {
+		return ok
+	}
+	logrus.Warnf("could not check md5 hashes for %s, comparing timestamps instead: %v", name, err)
+
+	localTime, err := dest.ModTime(name)
```

Comment:
Wandering, what is the reason to use `ModTime` in local FS? 

As posted in https://github.com/src-d/datasets/pull/58#issuecomment-388757968 - it seems that modification time is updated on my system for some reason, although the index file is old and corrupted.

Reply:
We use the mod time only for file system types (or files) that do not support MD5 hashes.
The main goal is to abstract the behavior so we can add other file system backends easily later on.